qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+
+        if carrier_freqs.ndim == 0:
+            carrier_freqs = Array([carrier_freqs])
+
+        if phases.ndim == 0:
+            phases = Array([phases])
 
-        return cls(
-            envelope=env_func,
+        if len(samples) == 1:
+            return DiscreteSignal(
+                dt=self.dt,
+                samples=samples[0],
+                start_time=self.start_time,
+                carrier_freq=carrier_freqs[0],
+                phase=phases[0],
+            )
+
+        return DiscreteSignalSum(
+            dt=self.dt,
+            samples=samples,
+            start_time=self.start_time,
             carrier_freqs=carrier_freqs,
             phases=phases,
-            drift_array=Array(drift_array),
         )
 
-    def envelope_value(self, t: float) -> Array:
-        """Evaluate the envelope.
 
-        Args:
-            t: time
+class SignalList:
+    """A list of ``Signal``s, with functionality for simultaneous evaluation."""
 
-        Returns:
-            Array: the signal envelope at time t
-        """
-        return self.envelope(t)
+    def __init__(self, signal_list: List[Signal]):
+        self.components = signal_list
 
-    def value(self, t: float) -> Array:
-        """Evaluate the full value of the VectorSignal.
+    def complex_value(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of complex value of components."""
+        return np.moveaxis(
+            Array([Array(sig.complex_value(t)).data for sig in self.components]), 0, -1
+        )
 
-        Args:
-            t (float): time
+    def __call__(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of all components."""
+        return np.moveaxis(Array([Array(sig(t)).data for sig in self.components]), 0, -1)
 
-        Returns:
-            Array: the value of the signal (including carrier frequencies)
-                      at time t
-        """
-        carrier_val = np.exp(t * self._im_angular_freqs + 1.0j * self.phases)
-        return self.envelope_value(t) * carrier_val
+    def flatten(self) -> "SignalList":
+        """Return a ``SignalList`` with each component flattened."""
+        flattened_list = []
+        for sig in self.components:
+            if isinstance(sig, SignalSum):
+                flattened_list.append(sig.flatten())
+            else:
+                flattened_list.append(sig)
 
-    def conjugate(self):
-        """Return a new VectorSignal that is the complex conjugate of self.
+        return SignalList(flattened_list)
 
-        Returns:
-            VectorSignal: the complex conjugate of self
-        """
-        return VectorSignal(
-            lambda t: np.conjugate(self.envelope_value(t)),
-            -self.carrier_freqs,
-            -self.phases,
-            np.conjugate(self.drift_array),
+    @property
+    def drift(self) -> Array:
+        """Return the 'drift' Array, i.e. the constant part of the ``SignalList``."""
```

Comment:
By `constant part` do you mean the sub-list of all constant signals? Can you clarify? So if understand this part it will allow us to conveniently separate drift and control parts of a Hamiltonian?

Reply:
I've updated this to try to be more clear.

And yes, you're correct, the ultimate use case is to be able to determine the "constant part" of a model (like a Hamiltonian).